File: 108. Convert Sorted Array to Binary Search Tree/recursion.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import math
class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        self.result = []
        if not nums:
            return self.result
        root = []
        def ConstructTree(root,nums):
            if not nums:
                return None
            if len(nums) == 1:
                root = TreeNode(nums[0])
                return root
            if len(nums) == 2:
                root = TreeNode(nums[1])
                root.left = TreeNode(nums[0])
                return root
            if len(nums) == 3:
                root = TreeNode(nums[1])
                root.left = TreeNode(nums[0])
                root.right = TreeNode(nums[2])
                return root
            root = TreeNode(nums[len(nums)//2])
            root.left = ConstructTree(root,nums[0:len(nums)//2])
            root.right = ConstructTree(root,nums[len(nums)//2+1:len(nums)])
            return root
        return ConstructTree(root,nums)

File: 108. Convert Sorted Array to Binary Search Tree/recursion.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode *ConstructTree(int *nums, int start, int end){
    struct TreeNode *root;
    root = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    int mid;
    if (start>end) return NULL;
    if (start==end){
        root->val = nums[start];
        root->left = NULL;
        root->right = NULL;
        return root;
    }
    mid = (end-start)/2 + start;
    root->val = nums[mid];
    root->left = ConstructTree(nums,start, mid-1);
    root->right = ConstructTree(nums,mid+1,end);
    return root;
}
struct TreeNode* sortedArrayToBST(int* nums, int numsSize) {
    return ConstructTree(nums, 0, numsSize-1);
}
